cares_reinforcement_learning/algorithm/policy/SACD.py

```diff
@@ -0,0 +1,198 @@
+"""
+Original Paper: https://arxiv.org/pdf/1910.07207
+Code based on: https://github.com/pranz24/pytorch-soft-actor-critic/blob/master/sac.py.
+
+This code runs automatic entropy tuning
+"""
+
+import copy
+import logging
+import os
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+
+import cares_reinforcement_learning.util.helpers as hlp
+from cares_reinforcement_learning.memory import MemoryBuffer
+
+
+class SACD:
+    def __init__(
+        self,
+        actor_network: torch.nn.Module,
+        critic_network: torch.nn.Module,
+        gamma: float,
+        tau: float,
+        reward_scale: float,
+        action_num: int,
+        actor_lr: float,
+        critic_lr: float,
+        alpha_lr: float,
+        device: torch.device,
+    ):
+        self.type = "discrete_policy"
```

Comment:
How does this get handled by gym_environment? is there a respective update? This should just stay as policy no?

Reply:
The normalization in the gym environment was breaking things so wanted to make a loop without normalization. Sorry about the horrible naming

Reply:
how is it breaking things? it shouldn't break anything

Reply:
In the discrete setting, `select_action_from_policy()` and `sample_action()` return a tensor containing the integer index of the action to enact. However, the action returned from `sample_action()` is normalised, while the action returned from `select_action_from_policy()` is denormalised. To make this work, we would need to configure some sort of redundant reversal of normalisation/denormalisation for one of these functions so that run_action_on_emulator() receives only the normalised or original action format instead of both.